### PR TITLE
Issue/4553 implement remove card event

### PR DESCRIPTION
**Pull request by malinajirka**
Parent issue: #4553

### Description

~This PR displays a snackbar when Stripe Terminal SDK emits "DisplayMessageRequested" event. The snackbar is shown on top of the dialog in order to make sure we don't show it behind the grey overlay. ~

~In general it is not recommended to show snackbars in dialogs, but I can't think of another solution. We "can't" show this event as a step (synchronous dialog) of the payment flow as the event can come in at any moment - even when the payment is being processed/captured or is completed (aka it is asynchronous - will be even more asynchronous in v2 of the SDK).~

Update: The PR update the hint on "Collect payment" step when the user is required to retry using the card. The PR also hardcodes "Remove Card" hint to process/capture payment steps to make sure we inform the user that they are safe to remove the card.

I decided to ignore "`Insert`/`Insert or swipe`/`swipe`" events, since the app is showing "Collect payment" screen with "Tap or insert to pay" static hint.

Note: This PR also fixes "onEmailActivityNotFound" error message as it was shown behind the grey overlay.

Note 2:Please feel free even encouraged to commit changes to this PR - especially with updates of the messages/hints.

### Testing instructions

- Use a store eligible for card present payments and an order in US$
1. Open a detail of an unpaid order in US$
2. Tap on the Collect Payment button
3. Connect to a reader
4. Wait until the app prompts you to tap/swipe/insert a card
5. Insert a card up-side-down (the chip facing ground)
6. Notice the hint gets updated to  "Retry inserting/tapping/swiping"
7. Put two contact less cards on top of each other and tap the reader (I used one stripe testing card and my actual card)
8. Notice the hint gets updated to  "Retry presenting a single card" message
9. Insert the testing card
10. Notice "Remove card" hint on processing/capturing steps

### Images/gif

Update video in [this comment](https://github.com/woocommerce/woocommerce-android/pull/4619#issuecomment-901677138)
https://user-images.githubusercontent.com/2261188/129387782-3ff94595-a478-4f72-b4d7-fa4132d04572.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by nbradbury** [comment]
@malinajirka I'm not sure what this snackbar is doing. Is it asking the user to remove the second card?

remove

I'm wondering if it would be better to show only the "Multiple cards detected" snackbar.

**Comment by nbradbury** [comment]
@malinajirka I know you said to feel free to modify the messaging, but I wanted to ask whether it's important to say "contactless" in the "Multiple contactless cards detected" message? I'd choose to remove that word, but wondered if there was a reason for that to be included.

Related, I was wondering about the "Retry using another read method" message. If that's asking to use a different reader, I'd want to shorten that to "Retry using another reader."

**Comment by malinajirka** [comment]
Thanks for the review!

> I'm not sure what this snackbar is doing. Is it asking the user to remove the second card?

"Remove the card" message is shown when everything succeeds and the "card reader" prompts the user to remove the inserted/tapped card since it's not necessary anymore.

> I'm wondering if it would be better to show only the "Multiple cards detected" snackbar. 

That snackbar is used in a different scenario - when the user taps the reader with more then one card (usually when they leave the card in their wallet and tap the reader with the whole wallet which contains more than one card).

>  to ask whether it's important to say "contactless" in the "Multiple contactless cards detected" message?

Technically it's not important - it's not possible two swipe or insert two cards, only tap with two cards, that's why I included the word "contactless". But I think we can remove it.

> Related, I was wondering about the "Retry using another read method" message. If that's asking to use a different reader, I'd want to shorten that to "Retry using another reader."

This message means "if you tapped the card, try to inserting/swiping it. If you tried inserting it, try swiping/tapping ...".

**Comment by nbradbury** [comment]
@jirka Let's just remove "contactless" from that one message. Once conflicts are resolved, I'm good to merge.

**Review by Garance91540 — CHANGES_REQUESTED** [review]
Hi there,

I'd prefer if we didn't overlay snackbars on top of modals, that's a pretty uncommon use case (snackbars are usually display full width at the bottom of the device), and it's also behaving like a modal on top of a modal which is not a recommended pattern.

Can we use the subtitle here to display the snackbar messaging? It's what it was initially intended for:

Screen Shot 2021-08-17 at 11 53 28 AM

I'm a bit confused by the "Multiple contactless cards detected" message: what are users supposed to do when multiple cards are detected? We could display some troubleshooting messaging in the subtitle text:

Screen Shot 2021-08-17 at 11 55 12 AM

**Comment by malinajirka** [comment]
Thanks for the review @Garance91540! I agree that using a snackbar for this purpose isn't great, but as I mentioned in the description I couldn't think of another solution (the message events are in a different channel than the payment flow events). Perhaps, we could:



**REMOVE_CARD**
1. We could ignore "remove card" event and always show "remove card" hint on "processing/capturing" steps. (downside is that when processing/capturing payment takes time it might be a bit confusing that the app keeps showing "remove card" even though the user has already removed it, but AFAIK we don't know if the card has already been removed or not)

or

2. We could completely ignore "remove card" event and not show anything - the card reader starts beeping when the card is still inserted after the payment is processed.




**RETRY_CARD, MULTIPLE_CONTACTLESS_CARDS_DETECTED,TRY_ANOTHER_READ_METHOD,TRY_ANOTHER_CARD**
1. We could update the hint when the event is received on the collecting payment step and ignore it on all the other steps.


I'll try to implement this tomorrow and we'll see how it works.

**Comment by Garance91540** [comment]
Hi @malinajirka, it's a bit hard for me to visualize, so let me know when you got a new round of iterations ;) (But yes, definitely without snackbars)

Here are the different messages in the designs for each step of the payment process, are you saying there's no way for us to know which steps we are at?

Screen Shot 2021-08-18 at 4 27 19 PM

**Comment by malinajirka** [comment]
> Here are the different messages in the designs for each step of the payment process, are you saying there's no way for us to know which steps we are at?

The app knows the current step (collect/process/capture/finished), but "remove/retry/etc" are not actual steps - they are just hints from the card reader that the app should somehow display. For example, when the reader emits "remove card" it doesn't stop the payment flow, it continues processing, capturing and finishing the payment. 
If the user leaves the card inserted in the reader, the "remove card" event will be emitted every 5 seconds (eg. even when the app is already showing "Payment successful" screen).

Either way, I think the only event that is tricky is actually the "remove card" event - all the other events require the user to take an action before the payment flow continues - eg. the user needs to retry tapping/inserting/swiping the card or retry using a different card. As we discussed in May, we theoretically could manually stop the payment flow when "remove_card" event is received and wait until the user removes the card - but complexity of this solution is imho not worth it.

I'll update the hint on the collect payment step and hardcode "remove card" hint on "process and capture payment" steps. I think that's a solution which gives us 95% with a low effort.

**Comment by malinajirka** [comment]
@Garance91540 This is the updated behavior, let me know what you think:

https://user-images.githubusercontent.com/2261188/130026447-1b724cb1-1396-44ad-b9d2-b2116c4153a6.mp4

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialogFragment.kt [hunk 1]
```diff
@@ -75,6 +78,9 @@ class CardReaderPaymentDialogFragment : DialogFragment(R.layout.card_reader_paym
                     )
                     is SendReceipt -> composeEmail(event.address, event.subject, event.content)
                     is ShowSnackbar -> uiMessageResolver.showSnack(event.message)
+                    is ShowSnackbarInDialog -> Snackbar.make(
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialogFragment.kt:81]: I kept the snackbar for the `onEmailActivityNotFound` scenario - it's a very edge case scenario anway.

**Comment by nbradbury** [comment]
> I think the only event that is tricky is actually the "remove card" event

If that event is tricky, do we even need to handle it? Regardless of whether the payment succeeds or fails, the user should know they should remove the card. Are there side effects to the user not removing the card when that event is received?

**Comment by malinajirka** [comment]
> If that event is tricky, do we even need to handle it? Regardless of whether the payment succeeds or fails, the user should know they should remove the card. Are there side effects to the user not removing the card when that event is received?

Actually, the current implementation doesn't handle it. We just ignore that event and we always show "Remove Card" hint on processing/capturing payment steps. AFAIK there are no side affects and as you said the user will remove the card eventually anyway. Moreover, the card reader makes a beep sound when the card is still inserted after the payment finishes.

**Comment by malinajirka** [comment]
@nbradbury The code is ready for another round - I'll just need to update the PR description.

**Comment by nbradbury** [comment]
@malinajirka  I'm still not wild about the "Retry using another read method" message because it doesn't describe the solution, but I can't come up with a succinct message that explains it better. But we could at least shorten it to just "Try another read method."

**Comment by malinajirka** [comment]
@nbradbury Actually the most up to date code uses `Retry tapping/inserting/swiping` instead of `Try another read method`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialogFragment.kt (hunk 1 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialogFragment.kt:81]: I think this is fine. As you said, it's very unlikely the device won't have an email app, and the message needs to be temporary in that unlikely situation.

**Review by nbradbury — APPROVED** [review]
All good! :shipit:

**Comment by malinajirka** [comment]
@Garance91540 Nick suggested to use `Multiple cards detected. Try again with a single card` message to convey both the reason and the next step. The other messages are as we agreed on during our call. Could you please review, if it look ok to you? Thanks!

| -- | -- | -- | -- |
| -- | -- | -- | -- |
| Screenshot_20210819-153222_WooCommerce (Dev) | Screenshot_20210819-135009_WooCommerce (Dev) | Screenshot_20210819-135421_WooCommerce (Dev) | Screenshot_20210819-135449_WooCommerce (Dev) |

**Comment by Garance91540** [comment]
> @Garance91540 Nick suggested to use Multiple cards detected. Try again with a single card message to convey both the reason and the next step. The other messages are as we agreed on during our call. Could you please review, if it look ok to you? Thanks!

That sounds right! 

One note, could we write "retry tapping, inserting or swiping your card" instead of "retry tapping/inserting/swiping"? Thanks!

**Comment by nbradbury** [comment]
@Garance91540 I've enabled auto-merge on this since it's awaiting your approval and we're in such different time zones.

**Review by Garance91540 — APPROVED** [review]

**Pull request merged**